MatrixSDK/ThreePidAdd/MX3PidAddManager.m

```diff
@@ -0,0 +1,706 @@
+        threePidAddSession.httpOperation = nil;
+        failure(error);
+    }];
+
+    return threePidAddSession;
+}
+
+- (void)finaliseAddPhoneNumberSession:(MX3PidAddSession*)threePidAddSession
+                            withToken:(NSString*)token
+                              success:(void (^)(void))success
+                              failure:(void (^)(NSError * _Nonnull))failure
+{
+    NSLog(@"[MX3PidAddManager] finaliseAddPhoneNumberSession: threePid: %@", threePidAddSession);
+
+    NSParameterAssert([threePidAddSession.medium isEqualToString:kMX3PIDMediumMSISDN]);
+
+    if (threePidAddSession.httpOperation || !threePidAddSession.sid)
+    {
+        NSError *error = [NSError errorWithDomain:MX3PidAddManagerErrorDomain
+                                             code:MX3PidAddManagerErrorDomainErrorInvalidParameters
+                                         userInfo:nil];
+        failure(error);
+        return;
+    }
+
+    MXWeakify(self);
+    threePidAddSession.httpOperation = [self submitValidationToken:token for3PidAddSession:threePidAddSession success:^{
+        MXStrongifyAndReturnIfNil(self);
+
+        MXHTTPOperation *operation;
+        if (self->doesServerSupportSeparateAddAndBind)
+        {
+            // https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928#2b-adding-a-3pid-to-hs-account-after-registration-post-msc2290
+            operation = [self->mxSession.matrixRestClient add3PIDOnlyWithSessionId:threePidAddSession.sid clientSecret:threePidAddSession.clientSecret success:^{
+
+                NSLog(@"[MX3PidAddManager] finaliseAddPhoneNumberSession: DONE: threePid: %@", threePidAddSession);
+
+                threePidAddSession.httpOperation = nil;
+                success();
+
+            } failure:^(NSError *error) {
+                threePidAddSession.httpOperation = nil;
+                failure(error);
+            }];
+        }
+        else
+        {
+            // https://gist.github.com/jryans/839a09bf0c5a70e2f36ed990d50ed928#2a-adding-a-3pid-to-hs-account-after-registration-pre-msc2290
+            operation = [self->mxSession.matrixRestClient add3PID:threePidAddSession.sid clientSecret:threePidAddSession.clientSecret bind:NO success:^{
+
+                NSLog(@"[MX3PidAddManager] finaliseAddPhoneNumberSession: DONE: threePid: %@", threePidAddSession);
+
+                threePidAddSession.httpOperation = nil;
+                success();
+
+            } failure:^(NSError *error) {
+                threePidAddSession.httpOperation = nil;
+                failure(error);
+            }];
+        }
+
+        if (operation)
+        {
+            [threePidAddSession.httpOperation mutateTo:operation];
+        }
+
+    } failure:^(NSError *error) {
+        threePidAddSession.httpOperation = nil;
+        failure(error);
+    }];
+}
+
+
+#pragma mark - Bind Email
+
+- (MX3PidAddSession*)startIdentityServerEmailSessionWithEmail:(NSString*)email
```

Comment:
Maybe rename `startIdentityServerBindEmailSessionWithEmail` to add more context like `startAddEmailSessionWithEmail` above.